### ISSUE TITLE
ImportError: No module named _ctypes

### ISSUE THREAD

**Issue by ConnorRules** [pantsbuild/pex#572]
It looks like the ctypes library is required by the pex bootstrapper. Our current python environment does not have that library. Is there any way to package ctypes in my pex build?

**Comment by jsirois** [comment]
There isn't. The actual ctypes library is part of the python stdlib, but it relies on the native library libffi. See here for some details: https://bugs.python.org/issue31652

This is unfortunate - we only use ctypes on linux and in the context of determining if the current linux host is manylinux1 compatible. We could spend some effort to make importing lazy such that we only trigger this error if you're actually running pex on linux _and_ you want manylinux support.

**Comment by ConnorRules** [comment]
Interesting. During initial debugging of this error I did install libffi to no avail. The only thing that worked was copying the ctypes.so and ctypes library directory into my python environment. 

Trying to find a good workaround. The python environment I'm working with is prepackaged with some enterprise software. So manipulating the modules in it is not preferred and for some reason they didn't pack ctypes with it even though that's standard.

Thanks for the reply

**Comment by jsirois** [comment]
> Interesting. During initial debugging of this error I did install libffi to no avail.

I should have been more clear: you need to have libffi(-dev) installed to have libffi headers available before you _build_ python. Unfortunately python will happily build without ctypes support. This is why you had to monkey with copying.

If you don't actually need manylinux1 support, then I think this issue stands as a bug/enhancement. We definitely could be lazy accessing the tiny bit of code that needs to consult ctypes. If you then built pexes with `--no-manylinux`, you should be good.

**Comment by Laevos** [comment]
Hey @jsirois, I've been working with @ConnorRules on this issue. A `--no-manylinux` build option in pants would definitely solve this for us. Alternatively, would it be at all possible to just wrap the import of ctypes in [glibc.py#L4](https://github.com/pantsbuild/pex/blob/1600d5b8d4a3a85fa8c6e17d0525da794a1f915a/pex/glibc.py#L4) in a try block? Thanks!

**Comment by jsirois** [comment]
To be clear `--no-manylinux` already exists. I'm just suggesting that its only valid for us to lazily import the ctypes code (or try/except as you suggest) if `--no-manylinux`.

**Comment by Laevos** [comment]
Right, okay, that makes sense. I saw and tried to use `--no-manylinux` with PEX, but ran into the same issue. Just trying to think about how we would send the status of `--no-manylinux` to the glibc module. As far as I can tell, `glibc.py` is only pulled in by `pep425tags.py`, which is in turn imported in `package.py`, `platforms.py`, `translator.py` and `interpreter.py`, so just being able to pass the flag to the glibc module seems ideal.

**Comment by jsirois** [comment]
There is no support for global flags so the information will need to be threaded through those callsites unfortunately. Is this something you're willing to work on @Laevos?

**Comment by Laevos** [comment]
Yeah, I'd be willing to spend some time on it. Right now to get things working we're just making a wrapper script that runs the PEX in a separate Python environment that does have ctypes available, but eventually it would be nice to switch to a solution that is less dependent on a hack like that.

**Comment by jsirois** [comment]
Excellent @Laevos - thank you! I've invited you to the contributors team. If you accept I can assign this issue to you. If not, that's fine too.

**Comment by Laevos** [comment]
Accepted, go right ahead, thanks!
